### PR TITLE
Add NTP based retention time

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Add NTP based retention time

## Motivation and Context
To disallow depending on clock drifts on host system

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation needed
- [ ] Unit tests needed
- [ ] Functional tests needed (If yes, add [mint](https://github.com/minio/mint) PR # here: )

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :x: [more...](#af8ed6ec-b86f-480b-9618-1dbe6351a410) |
| mint-large-bucket.sh | :x: [more...](#5035d6c3-3180-4430-ae2d-c7f7ba73cb86) |
<a name="5035d6c3-3180-4430-ae2d-c7f7ba73cb86"></a>
### 8548-e2d8a8f/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.61:31251
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp cbc558748bfb:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 6 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 47 seconds
(4/14) Running aws-sdk-ruby tests ... done in 11 seconds
(5/14) Running awscli tests ... done in 2 minutes and 22 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 1 minutes and 2 seconds
(8/14) Running minio-dotnet tests ... done in 1 minutes and 5 seconds
(9/14) Running minio-go tests ... FAILED in 4 minutes and 6 seconds
{
  "args": {},
  "duration": 108934,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 8 out of 14 tests successfully.
```
<a name="af8ed6ec-b86f-480b-9618-1dbe6351a410"></a>
### 8548-e2d8a8f/mint-compression-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.52:30364
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 127c411e19ab:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 6 minutes and 5 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-1t0vxczqwxiluk",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 364901,
  "error": "XMinioServerNotInitialized: Server not initialized, please try again.\n\tstatus code: 503, request id: 15D90FA3C809D15A, host id: ",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```
Error: No such image: minio/minio:8548-e2d8a8f

File: cmd/object-lock.go [hunk 1]
```diff
@@ -63,6 +65,8 @@ var (
 	errUnknownWORMModeDirective    = errors.New("unknown WORM mode directive")
 	errObjectLockMissingContentMD5 = errors.New("Content-MD5 HTTP header is required for Put Object requests with Object Lock parameters")
 	errObjectLockInvalidHeaders    = errors.New("x-amz-object-lock-retain-until-date and x-amz-object-lock-mode must both be supplied")
+
+	ntpServer = env.Get("MINIO_NTP_SERVER", "0.us.pool.ntp.org")
```
Comment (kannappanr) [inline cmd/object-lock.go]: Can this be turned off?

Comment (harshavardhana) [reply cmd/object-lock.go]: It can't be turned off @kannappanr

Comment (klauspost) [reply cmd/object-lock.go]: Why not?

File: cmd/utils.go [hunk 2]
```diff
@@ -314,6 +315,17 @@ func UTCNow() time.Time {
...
 
+// UTCNowNTP - is similar in functionality to UTCNow()
+// but only used when we do not wish to rely on system
+// time.
+func UTCNowNTP() (time.Time, error) {
+	// ntp server is disabled
+	if ntpServer == "" {
+		return UTCNow(), nil
+	}
+	return ntp.Time(ntpServer)
```
Comment (klauspost) [inline cmd/utils.go:326]: There is a default timeout of 5 seconds and NTP is using UPD, so if it unable to get a reply that will always add 5 seconds to the operation.

Enabling it by default seems to just be asking for trouble. I'd propose it to be opt-in, so users will also explicitly have to find a close-by NTP server.

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#ad85826d-06c2-4571-8b6c-20cab59b0bfe) |
| mint-fs.sh | :x: [more...](#1f5b3293-9139-4313-b40b-862c1fccb4d3) |
<a name="ad85826d-06c2-4571-8b6c-20cab59b0bfe"></a>
### 8548-901aeac/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.54:32310
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp f0120a4daaea:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 1 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 1 minutes and 0 seconds
(4/14) Running aws-sdk-ruby tests ... done in 8 seconds
(5/14) Running awscli tests ... done in 2 minutes and 21 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 1 minutes and 20 seconds
(8/14) Running minio-dotnet tests ... done in 1 minutes and 21 seconds
(9/14) Running minio-go tests ... FAILED in 4 minutes and 26 seconds
{
  "args": {},
  "duration": 116844,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 8 out of 14 tests successfully.
```
<a name="1f5b3293-9139-4313-b40b-862c1fccb4d3"></a>
### 8548-901aeac/mint-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.56:31643
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 4bc9af367c9c:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 0 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 41 seconds
(4/14) Running aws-sdk-ruby tests ... done in 2 seconds
(5/14) Running awscli tests ... done in 53 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 28 seconds
(8/14) Running minio-dotnet tests ... done in 24 seconds
(9/14) Running minio-go tests ... done in 41 seconds
(10/14) Running minio-java tests ... done in 15 seconds
(11/14) Running minio-js tests ... FAILED in 18 seconds
{
  "name": "minio-js",
  "function": "fGetObject(bucketName, objectName, filePath)",
  "args": "bucketName:minio-js-test-093f6c1a-7afb-401f-8fb9-9175e581c2fa, objectName:datafile-65-MB, filePath:/tmp/datafile-65-MB.download",
  "duration": 798,
  "status": "FAIL",
  "error": "Error: Size mismatch between downloaded file and the object at /mint/run/core/minio-js/node_modules/minio/dist/main/minio.js:965:12 at nextTask (node_modules/minio/node_modules/async/dist/async.js:4578:13) at next (node_modules/minio/node_modules/async/dist/async.js:4586:13) at /mint/run/core/minio-js/node_modules/minio/node_modules/async/dist/async.js:325:20 at FSReqWrap.oncomplete (fs.js:123:15)"
}

Executed 10 out of 14 tests successfully.
```
Error: No such image: minio/minio:8548-901aeac

File: cmd/object-lock.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/object-lock.go]: This is meant for retention alone @klauspost it is assumed that we can't rely on system clock

File: cmd/utils.go (hunk 2 above)
Comment (harshavardhana) [reply cmd/utils.go:326]: This is a proposed PR not a final one @klauspost

Comment (harshavardhana) [reply cmd/utils.go:326]: Made it opt-in anyways @klauspost

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#7b5aab73-7e1d-44e5-bd90-709a0a7b6a05) |
<a name="7b5aab73-7e1d-44e5-bd90-709a0a7b6a05"></a>
### 8548-32855d9/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.51:31473
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp d19a9715429a:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 2 seconds
(2/14) Running aws-sdk-java tests ... done in 2 seconds
(3/14) Running aws-sdk-php tests ... done in 51 seconds
(4/14) Running aws-sdk-ruby tests ... done in 10 seconds
(5/14) Running awscli tests ... done in 2 minutes and 21 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 1 minutes and 15 seconds
(8/14) Running minio-dotnet tests ... done in 1 minutes and 5 seconds
(9/14) Running minio-go tests ... FAILED in 3 minutes and 26 seconds
{
  "args": {},
  "duration": 69185,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 8 out of 14 tests successfully.
```
Error: No such image: minio/minio:8548-32855d9

**Review by kannappanr — APPROVED** [review]
LGTM

**Review by Praveenrajmani — APPROVED** [review]
LGTM

**Pull request merged**